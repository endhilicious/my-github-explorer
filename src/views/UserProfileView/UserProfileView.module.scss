.userProfile {
  max-width: 800px;
  margin: 0 auto;
  padding: 1rem;
  main {
    margin-top: 2rem;
  }
  
  h1 {
    font-size: 1.5rem;
  }
}
.selectedImage {
  width: 50px;
  border-radius: 50%;
}
.cardHeaderSelected {
  display: flex;
  justify-content: space-between;
  padding: 10px;
  box-shadow: 3px 5px 6px #eaeaea;
  border: 1px solid #eaeaea;
  gap: 8px;
  & > div:last-child {
    display: flex;
    align-items: center;
    gap: 6px;
    & svg {
      cursor: pointer;
    }
  }
}