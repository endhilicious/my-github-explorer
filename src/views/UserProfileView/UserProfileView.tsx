import React, { useEffect } from 'react';
import { useRouter } from 'next/router';
import cx from 'classnames';

import Header from '@/components/Header';
import CloseIcon from '@/common/Icons/CloseIcon';
import StarIcon from '@/common/Icons/StarIcon';

import { useLoading } from '@/context/useLoading';
import { useToast } from '@/context/useToast';
import { useGithubData } from '@/context/useGithubData';
import { getUserDetail, getUserRepositories } from '@/utils/api';

import styles from './UserProfileView.module.scss';

const UserProfileView: React.FC = () => {
  const router = useRouter();
  const { username } = router.query;

  const { setLoading } = useLoading();
  const { showToast } = useToast();

  const { state: { selectedUser, repos }, dispatch } = useGithubData();

  const fetchUserData = async () => {
    try {
      setLoading(true)

      const userDetail = await getUserDetail(username as string);
      const userRepo = await getUserRepositories(username as string);

      dispatch({ type: 'SELECTED_USER', payload: userDetail });
      dispatch({ type: 'ADD_REPOS', payload: userRepo });

      showToast('Success fetching user repositories', 'success');
    } catch (error) {
      showToast('Error fetching user repositories', 'danger');

      console.error('Error fetching user repositories:', error);
    } finally {
      setLoading(false)
    }
  }

  useEffect(() => {
    if (username) fetchUserData();
  }, [username]);

  const handleCloseSelectedItem = () => {
    router.back();
  }

  return (
    <div className={styles.userProfile}>
      <Header />
      <main>
        <h1>{username}'s Repositories</h1>
        <div>
          {selectedUser && (
            <div className={styles.cardHeaderSelected}>
                <div>
                  <img className={styles.selectedImage} src={selectedUser?.avatar_url} alt={`${selectedUser?.login}'s avatar`} />
                </div>
                <div>
                  <h3>
                    {selectedUser?.login}
                  </h3>
                  <div onClick={handleCloseSelectedItem}>
                    <CloseIcon />
                  </div>
                </div>
            </div>
          )}
          {repos.map((newValue: any) => {
            return (
              <div className={cx(styles.cardHeaderSelected, styles.cardHeaderSelectedRepo)} style={{ margin: '10px 0'}}>
                <div>
                  <div>
                    <b>{newValue.name}</b>
                  </div>
                  <div>
                    <small>{newValue.description}</small>
                  </div>
                </div>
                <div>
                  <span>{newValue.stargazers_count}</span>
                  <span><StarIcon /></span>
                </div>
              </div>
            )
          })}
        </div>
      </main>
    </div>
  );
};

export default UserProfileView;
