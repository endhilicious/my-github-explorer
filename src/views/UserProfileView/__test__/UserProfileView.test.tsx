import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { useRouter } from 'next/router';
import UserProfileView from '../UserProfileView';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

jest.mock('../../../context/useLoading', () => ({
  useLoading: jest.fn(() => ({ setLoading: jest.fn() })),
}));

jest.mock('../../../context/useToast', () => ({
  useToast: jest.fn(() => ({ showToast: jest.fn() })),
}));

jest.mock('../../../context/useGithubData', () => ({
  useGithubData: jest.fn(() => ({
    state: {
      selectedUser: {
        login: 'testuser',
        avatar_url: 'https://example.com/avatar.png',
      },
      repos: [
        {
          name: 'repo1',
          description: 'Description 1',
          stargazers_count: 10,
        },
        {
          name: 'repo2',
          description: 'Description 2',
          stargazers_count: 20,
        },
      ],
    },
    dispatch: jest.fn(),
  })),
}));

jest.mock('../../../utils/api', () => ({
  getUserDetail: jest.fn(() => Promise.resolve({})),
  getUserRepositories: jest.fn(() => Promise.resolve([])),
}));

describe('UserProfileView Component', () => {
  it('renders UserProfileView component correctly', () => {
    (useRouter as any).mockReturnValue({ query: { username: 'testuser' } });

    const { asFragment } = render(<UserProfileView />);
    expect(asFragment()).toMatchSnapshot();
  });
});
