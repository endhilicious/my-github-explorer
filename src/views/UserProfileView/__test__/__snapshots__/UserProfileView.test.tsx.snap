// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileView Component renders UserProfileView component correctly 1`] = `
<DocumentFragment>
  <div
    class="userProfile"
  >
    <header
      class="header"
    >
      <h1
        data-testid="header-text"
      >
        Github Repositories Explorer
      </h1>
    </header>
    <main>
      <h1>
        testuser's Repositories
      </h1>
      <div>
        <div
          class="cardHeaderSelected"
        >
          <div>
            <img
              alt="testuser's avatar"
              class="selectedImage"
              src="https://example.com/avatar.png"
            />
          </div>
          <div>
            <h3>
              testuser
            </h3>
            <div>
              <svg
                fill="none"
                height="24"
                viewBox="0 0 48 48"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M10.5858 37.414C10.9768 37.805 11.4878 38 11.9998 38C12.5118 38 13.0228 37.805 13.4138 37.414L24 26.8279L34.5862 37.414C34.9772 37.805 35.4882 38 36.0002 38C36.5122 38 37.0232 37.805 37.4142 37.414C38.1953 36.633 38.1953 35.367 37.4142 34.586L26.8281 23.9999L37.4142 13.4138C38.1953 12.6328 38.1953 11.3668 37.4142 10.5858C36.6332 9.80475 35.3672 9.80475 34.5862 10.5858L24 21.1718L13.4138 10.5858C12.6328 9.80475 11.3668 9.80475 10.5858 10.5858C9.80475 11.3668 9.80475 12.6328 10.5858 13.4138L21.1719 23.9999L10.5858 34.586C9.80475 35.367 9.80475 36.633 10.5858 37.414Z"
                  fill="#212121"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="cardHeaderSelected cardHeaderSelectedRepo"
          style="margin: 10px 0px;"
        >
          <div>
            <div>
              <b>
                repo1
              </b>
            </div>
            <div>
              <small>
                Description 1
              </small>
            </div>
          </div>
          <div>
            <span>
              10
            </span>
            <span>
              <svg
                fill="#000"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          class="cardHeaderSelected cardHeaderSelectedRepo"
          style="margin: 10px 0px;"
        >
          <div>
            <div>
              <b>
                repo2
              </b>
            </div>
            <div>
              <small>
                Description 2
              </small>
            </div>
          </div>
          <div>
            <span>
              20
            </span>
            <span>
              <svg
                fill="#000"
                height="18"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polygon
                  points="12 2 15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
